plugins/catalog-backend-module-incremental-ingestion/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -546,12 +560,26 @@ export class IncrementalIngestionDatabaseManager {
    * @param entities - DeferredEntity[]
    */
   async createMarkEntities(markId: string, entities: DeferredEntity[]) {
+    const refs = entities.map(e => stringifyEntityRef(e.entity));
+
     await this.client.transaction(async tx => {
+      const existingRefs = (
+        await tx<{ ref: string }>('ingestion_mark_entities')
+          .select('ref')
+          .whereIn('ref', refs)
+      ).map(e => e.ref);
+
+      const newRefs = refs.filter(e => !existingRefs.includes(e));
```

Comment:
Just a small note that this is `O(N^2)` and for large numbers of items it might have been sensible to make `existingRefs` a `Set` instead.

Reply:
This is a good reminder to use Set when possible. The whole goal of incremental providers is to improve performance and efficiency, so I've gone ahead and implemented this.